### PR TITLE
Enable metadata service to track the events from a subset of namespaces

### PR DESCRIPTION
**Pull request by MrAta**
Summary: Enable metadata service to track the events from a subset of namespaces when desired, instead of tracking the events from the entire cluster.

Relevant Issues: #1364

Type of change: /kind feature

Test Plan: Added unit tests for the new behaviour. @JamesMBartlett tested by skaffolding a dev vizier version to a cluster both with and without `metadata_namespaces` selected, and both behaved as expected.

### CODE REVIEW

File: src/vizier/services/metadata/metadata_server.go [hunk 1]
```diff
@@ -235,7 +236,11 @@ func main() {
 	updateCh := make(chan *k8smeta.K8sResourceMessage)
 	mdh := k8smeta.NewHandler(updateCh, k8sMds, k8sMds, nc)
 
-	k8sMc, err := k8smeta.NewController(updateCh)
+	namespaces := []string{v1.NamespaceAll}
+	if viper.IsSet("METADATA_NAMESPACES") {
+		namespaces = strings.Split(viper.GetString("METADATA_NAMESPACES"), ",")
+	}
```
Comment (vihangm) [inline src/vizier/services/metadata/metadata_server.go:242]: Can you explicitly declare this as a pflag in `init` and use that instead? Then you can make it a `StringSlice`.

File: src/vizier/services/metadata/metadata_server.go [hunk 2]
```diff
@@ -36,6 +36,7 @@ import (
 	"go.etcd.io/etcd/client/pkg/v3/transport"
 	clientv3 "go.etcd.io/etcd/client/v3"
 	"google.golang.org/grpc"
+	v1 "k8s.io/api/core/v1"
```
Comment (vihangm) [inline src/vizier/services/metadata/metadata_server.go:39]: Looks like you need to run `gazelle` to update the build files to add this dep.
If you have `bazel` and the golang toolchain setup `make gazelle` at top of tree should do the trick.

**Review by vihangm — CHANGES_REQUESTED** [review]

File: src/vizier/services/metadata/metadata_server.go [hunk 3]
```diff
@@ -73,13 +74,15 @@ func init() {
...
 	pflag.Bool("use_etcd_operator", false, "Whether the etcd operator should be used instead of the persistent version.")
+	pflag.StringSlice("metadata_namespaces", []string{v1.NamespaceAll}, "The list of namespaces to watch for metadata.")
 
 	// Metadata flags are set using the env vars in pl-cluster-config.
 	// We historically set PL_ETCD_OPERATOR_ENABLED but not PL_USE_ETCD_OPERATOR in the configmap.
 	// We also don't have a clean way to update  configmaps for existing deploys.
 	// So instead just map PL_ETCD_OPERATOR_ENABLED to use_etcd_operator to make it work.
 	// TODO: We should clean this up in the future and make these flags consistent.
 	viper.BindEnv("use_etcd_operator", "PL_ETCD_OPERATOR_ENABLED")
+	viper.BindEnv("metadata_namespaces")
```
Comment (vihangm) [inline src/vizier/services/metadata/metadata_server.go:85]: This isn't necessary. Our shared service setup code will bind all flags to corresponding env vars with a `PL_` prefix.

File: src/vizier/services/metadata/controllers/k8smeta/k8s_metadata_utils.go [hunk 4]
```diff
@@ -51,38 +51,46 @@ func (i *informerWatcher) send(msg *K8sResourceMessage, et watch.EventType) {
...
-	// We initialize ch with the current nodes to handle cold start race conditions.
+
+	for _, ns := range namespaces {
+		factory := informers.NewSharedInformerFactoryWithOptions(clientset, 12*time.Hour, informers.WithNamespace(ns))
+		inf := factory.Core().V1().Pods().Informer()
+		iw.informers = append(iw.informers, inf)
+	}
+
+	// We initialize ch with the current Pods to handle cold start race conditions.
 	list, err := clientset.CoreV1().Pods(v1.NamespaceAll).List(context.Background(), metav1.ListOptions{})
```
Comment (vihangm) [inline src/vizier/services/metadata/controllers/k8smeta/k8s_metadata_utils.go:94]: the namespaces here need to also be restricted.

**Comment by vihangm** [comment]
This looks mostly ok, but I think #1595 will actually make this easier by allowing us to filter in only one location (i.e. at the SharedIndexInformerFactory creation).

Also we should see if it's possible to add some testing coverage to ensure that the namespace restriction works as expected.

**Comment by MrAta** [comment]
Thanks for the insights, @vihangm!
Re #1595, either of the PRs will need a rebase and conflict resolution once the other one is merged :D I'm not sure which way is more convenient, up to you.

Re the test case, I just added a simple test case for Pod creation only. Please let me know if it sounds right to you and I can add other resource types and events too.

**Comment by vihangm** [comment]
Test case looks good. Adding a little bit of simple coverage on all resources types would be nice.

**Comment by MrAta** [comment]
@vihangm add more test cases for other resource types.

**Review by JamesMBartlett — APPROVED** [review]
LGTM, discussed with @vihangm who's out at the moment and he said its good to go from his end. So if we can get the merge conflicts fixed we should be good to go. Sorry for the delay

**Comment by vihangm** [comment]
Also can you update the test plan in the PR since this now has unit tests (and also should be tested with a dev deploy of vizier before merging).

**Comment by JamesMBartlett** [comment]
I tested this out on a cluster, both with all namespaces and with a subset of namespaces and both seemed to work as expected.

**Review by JamesMBartlett — APPROVED** [review]

**Pull request merged**